cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -196,6 +196,11 @@ static inline Future<std::shared_ptr<csv::StreamingReader>> OpenReaderAsync(
   auto reader_fut = DeferNotOk(input->io_context().executor()->Submit(
       [=]() -> Future<std::shared_ptr<csv::StreamingReader>> {
         ARROW_ASSIGN_OR_RAISE(auto first_block, input->Peek(reader_options.block_size));
+        auto size = first_block.length();
+        const uint8_t* data = reinterpret_cast<const uint8_t*>(first_block.data());
+        ARROW_ASSIGN_OR_RAISE(auto data_no_bom, util::SkipUTF8BOM(data, size));
+        size = size - static_cast<uint32_t>(data_no_bom - data);
+        first_block = util::string_view(reinterpret_cast<const char*>(data_no_bom), size);
```

Comment:
@pitrou, according to your comment, I've moved the skip bom code to first block only.

Reply:
Ok, but I still don't understand why you need BOM-handling code in the datasets layer while the CSV reader already has such code.

Also, I see no tests being added here.

Reply:
I see the code in file_csv runs before the csv reader. And as the comment says it's to `Grab the first block and use it to determine the schema and create a reader.`

Reply:
Hmm, I see, thank you! But `SkipUTF8BOM` is already called in `GetColumnNames` above. Can we just fix `GetColumnNames` so that it calls `SkipUTF8BOM` at the right point?

Reply:
Thank you @pitrou. I've moved it to `GetColumnNames`.